src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs

```diff
@@ -125,6 +125,77 @@ private static void SendToRecvFromAsync_Datagram_UDP(IPAddress leftAddress, IPAd
             }
         }
 
+        private static void SendToRecvFromAsync_UdpClient_Datagram_UDP(IPAddress leftAddress, IPAddress rightAddress)
+        {
+            const int DatagramSize = 256;
+            const int DatagramsToSend = 256;
+            const int AckTimeout = 1000;
+            const int TestTimeout = 30000;
+
+            var left = new UdpClient(new IPEndPoint(leftAddress, 0));
+            var right = new UdpClient(new IPEndPoint(rightAddress, 0));
+
+            var leftEndpoint = (IPEndPoint)left.Client.LocalEndPoint;
+            var rightEndpoint = (IPEndPoint)right.Client.LocalEndPoint;
+
+            var receiverAck = new ManualResetEventSlim();
+            var senderAck = new ManualResetEventSlim();
+
+            var receivedChecksums = new uint?[DatagramsToSend];
+            int receivedDatagrams = 0;
+
+            Task receiverTask = Task.Run(async () =>
+            {
+                for (; receivedDatagrams < DatagramsToSend; receivedDatagrams++)
+                {
+                    UdpReceiveResult result = await left.ReceiveAsync();
+
+                    receiverAck.Set();
+                    Assert.True(senderAck.Wait(AckTimeout));
+                    senderAck.Reset();
+
+                    Assert.Equal(DatagramSize, result.Buffer.Length);
+                    Assert.Equal(rightEndpoint, result.RemoteEndPoint);
+
+                    int datagramId = (int)result.Buffer[0];
+                    Assert.Null(receivedChecksums[datagramId]);
+
+                    receivedChecksums[datagramId] = Fletcher32.Checksum(result.Buffer, 0, result.Buffer.Length);
+                }
+            });
+
+            var sentChecksums = new uint[DatagramsToSend];
+            int sentDatagrams = 0;
+
+            Task senderTask = Task.Run(async () =>
+            {
+                var random = new Random();
+                var sendBuffer = new byte[DatagramSize];
+
+                for (; sentDatagrams < DatagramsToSend; sentDatagrams++)
+                {
+                    random.NextBytes(sendBuffer);
+                    sendBuffer[0] = (byte)sentDatagrams;
+
+                    int sent = await right.SendAsync(sendBuffer, DatagramSize, leftEndpoint);
+
+                    Assert.True(receiverAck.Wait(AckTimeout));
+                    receiverAck.Reset();
+                    senderAck.Set();
+
+                    Assert.Equal(DatagramSize, sent);
+                    sentChecksums[sentDatagrams] = Fletcher32.Checksum(sendBuffer, 0, sent);
+                }
+            });
+
+            Assert.True(Task.WaitAll(new[] { receiverTask, senderTask }, TestTimeout));
+            for (int i = 0; i < DatagramsToSend; i++)
+            {
+                Assert.NotNull(receivedChecksums[i]);
+                Assert.Equal(sentChecksums[i], (uint)receivedChecksums[i]);
```

Comment:
> An improvement would be to add a simple matching protocol and allow for missed receives.

There is a matching protocol: the first byte of each datagram is a sequence number. There is not currently any accommodation for dropped packets, however. This is something I'm thinking about at the moment.